pkg/hubble/monitor/consumer.go

```diff
@@ -57,73 +59,55 @@ func NewConsumer(observer Observer) monitorConsumer.MonitorConsumer {
-// counter to the observer. If it succeeds to enqueue a notification, it
-// resets the counter. Returns a boolean indicating whether the notification
-// has been successfully sent.
-func (c *consumer) sendNumLostEvents() bool {
-	c.lostLock.Lock()
-	defer c.lostLock.Unlock()
-	// check again, in case multiple
-	// routines contended the lock
-	if c.numEventsLost == 0 {
-		return true
-	}
+// NotifyAgentEvent implements monitorConsumer.MonitorConsumer.
+func (c *consumer) NotifyAgentEvent(typ int, message any) {
+	c.sendEvent(func() any {
+		return &observerTypes.AgentEvent{
+			Type:    typ,
+			Message: message,
+		}
+	})
+}
 
-	if c.cachedLostNotification == nil {
-		c.cachedLostNotification = c.newEvent(func() any {
-			return &observerTypes.LostEvent{
-				Source:        observerTypes.LostEventSourceEventsQueue,
-				NumLostEvents: c.numEventsLost,
-			}
-		})
-	} else {
-		c.cachedLostNotification.Timestamp = time.Now()
-		c.cachedLostNotification.Payload.(*observerTypes.LostEvent).NumLostEvents = c.numEventsLost
-	}
+// NotifyPerfEvent implements monitorConsumer.MonitorConsumer.
+func (c *consumer) NotifyPerfEvent(data []byte, cpu int) {
+	c.sendEvent(func() any {
+		return &observerTypes.PerfEvent{
+			Data: data,
+			CPU:  cpu,
+		}
+	})
+}
 
-	select {
-	case c.observer.GetEventsChannel() <- c.cachedLostNotification:
-		// We now now safely reset the counter, as at this point have
-		// successfully notified the observer about the amount of events
-		// that were lost since the previous LostEvent message. Similarly,
-		// we reset the cached notification, so that a new one is created
-		// the next time.
-		c.numEventsLost = 0
-		c.cachedLostNotification = nil
-		return true
-	default:
-		// We do not need to bump the numEventsLost counter here, as we will
-		// try to send a new LostEvent notification again during the next
-		// invocation of sendEvent
-		return false
-	}
+// NotifyPerfEventLost implements monitorConsumer.MonitorConsumer.
+func (c *consumer) NotifyPerfEventLost(numLostEvents uint64, cpu int) {
+	c.sendEvent(func() any {
+		return &observerTypes.LostEvent{
+			Source:        observerTypes.LostEventSourcePerfRingBuffer,
+			NumLostEvents: numLostEvents,
+			CPU:           cpu,
+		}
+	})
+	c.metricLostPerfEvents.Inc()
 }
 
 // sendEvent enqueues an event in the observer. If this is not possible, it
-// keeps a counter of lost events, which it will regularly try to send to the
-// observer as well
+// keeps a counter of lost events, which it will try to send at most once per
+// configured interval, and on every call to sendEvent until it succeeds.
 func (c *consumer) sendEvent(payloader func() any) {
-	if c.numEventsLost > 0 {
-		if !c.sendNumLostEvents() {
-			// We just failed sending the lost notification, hence it doesn't
-			// make sense to try and send the actual event, as we'll most
-			// likely fail as well.
-			c.countDroppedEvent()
-			return
-		}
-	}
+	now := time.Now()
+	c.trySendLostEvent(now)
```

Comment:
I changed this to always try to insert the event afterward regardless of whether we could send the lost event.

My reasoning is that there is nothing that guarantee we won't succeed on the next send to the channel, and since we will retry to send the lost event on each send until it succeeds, we should probably also try to send the original event and have them both race the channel. It's more important that the original event wins, even if it means trying to send twice per send calls.